akka-projection-grpc/src/main/scala/akka/projection/grpc/internal/EventsBySlicesFirehose.scala

```diff
@@ -89,36 +102,135 @@ object EventsBySlicesFirehose extends ExtensionId[EventsBySlicesFirehose] {
+
+        val slowConsumers = consumerTrackingValues.collect {
+          case t
+              if t.firehoseOnly &&
+              isDurationGreaterThan(t.timestamp, fastestConsumer.timestamp, JDuration.ofSeconds(5)) // FIXME config
+              =>
+            t.consumerId -> t
+        }.toMap
+
+        val changedConsumerTrackingValues = consumerTrackingValues.flatMap { tracking =>
+          if (slowConsumers.contains(tracking.consumerId)) {
+            if (tracking.slowConsumerCandidate.isDefined)
+              None // keep original
+            else
+              Some(tracking.copy(slowConsumerCandidate = Some(now)))
+          } else if (tracking.slowConsumerCandidate.isDefined) {
+            Some(tracking.copy(slowConsumerCandidate = None)) // not slow any more
+          } else {
+            None
+          }
+        }
+
+        changedConsumerTrackingValues.foreach { tracking =>
+          consumerTracking.merge(
+            tracking.consumerId,
+            tracking,
+            (existing, _) => existing.copy(slowConsumerCandidate = tracking.slowConsumerCandidate))
+        }
+
+        val newConsumerTrackingValues = consumerTracking.values.iterator.asScala.toVector
+
+        // FIXME trace
+        if (log.isDebugEnabled && newConsumerTrackingValues.iterator.map(_.timestamp).toSet.size > 1) {
+          newConsumerTrackingValues.foreach { tracking =>
+            val diffFastest = fastestConsumer.timestamp.toEpochMilli - tracking.timestamp.toEpochMilli
+            val diffFastestStr =
+              if (diffFastest > 0) s"behind fastest [$diffFastest] ms"
+              else if (diffFastest < 0) s"ahead of fastest [$diffFastest] ms" // not possible
+              else "same as fastest"
+            val diffSlowest = slowestConsumer.timestamp.toEpochMilli - tracking.timestamp.toEpochMilli
+            val diffSlowestStr =
+              if (diffSlowest > 0) s"behind slowest [$diffSlowest] ms" // not possible
+              else if (diffSlowest < 0) s"ahead of slowest [${-diffSlowest}] ms"
+              else "same as slowest"
+            log.debug(
+              "Firehose entityType [{}] sliceRange [{}] consumer [{}], {}, {}, firehoseOnly [{}]",
+              entityType,
+              sliceRange,
+              tracking.consumerId,
+              diffFastestStr,
+              diffSlowestStr,
+              tracking.firehoseOnly)
+          }
+        }
+
+        val firehoseConsumerCount = newConsumerTrackingValues.count(_.firehoseOnly)
+        val confirmedSlowConsumers = newConsumerTrackingValues.filter { tracking =>
+          tracking.slowConsumerCandidate match {
+            case None => false
+            case Some(detectedTimestamp) =>
+              isDurationGreaterThan(detectedTimestamp, now, JDuration.ofSeconds(3))
+          }
+        }
+
+        if (confirmedSlowConsumers.nonEmpty && confirmedSlowConsumers.size < firehoseConsumerCount) {
```

Comment:
The scenario where all firehoseConsumers are slow is not possible, right?

Reply:
added a fixme for now